crates/ra_hir/src/expr/validation.rs

```diff
@@ -29,11 +33,23 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
 
     pub(crate) fn validate_body(&mut self, db: &impl HirDatabase) {
         let body = self.func.body(db);
+
+        // The final expr in the function body is the whole body,
+        // so the expression being returned is the penultimate expr.
+        let mut penultimate_expr = None;
```

Comment:
Rather than rely on expressions order, I think we should take `body_expr` out of `Body`, match it to get an `Expr::Block` and use `tail` from that.